### PR TITLE
Add `actor.__ray_kill__()` to terminate actors immediately

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

If an actor is hanging or an application is shutting down, sometimes you need to terminate actors immediately without finishing outstanding work.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -1065,6 +1072,24 @@ void CoreWorker::HandleGetObjectStatus(const rpc::GetObjectStatusRequest &reques
...
+    std::ostringstream stream;
+    stream << "Mismatched ActorID: ignoring ForceKillActor for previous actor "
+           << intended_actor_id
+           << ", current actor ID: " << worker_context_.GetCurrentActorID();
+    auto msg = stream.str();
+    RAY_LOG(ERROR) << msg;
+    send_reply_callback(Status::Invalid(msg), nullptr, nullptr);
+    return;
+  }
+  RAY_LOG(INFO) << "Got ForceKillActor, shutting down...";
```
Comment (richardliaw) [inline src/ray/core_worker/core_worker.cc]: can you log the actor ID too?

File: src/ray/core_worker/transport/direct_actor_transport.cc [hunk 2]
```diff
@@ -101,6 +120,15 @@ void CoreWorkerDirectActorTaskSubmitter::DisconnectActor(const ActorID &actor_id
 void CoreWorkerDirectActorTaskSubmitter::SendPendingTasks(const ActorID &actor_id) {
   auto &client = rpc_clients_[actor_id];
   RAY_CHECK(client);
+  // Check if there is a pending force kill. If there is, send it and disconnec the
```
Comment (richardliaw) [inline src/ray/core_worker/transport/direct_actor_transport.cc]:
```suggestion
  // Check if there is a pending force kill. If there is, send it and disconnet the
```

File: python/ray/actor.py [hunk 3]
```diff
@@ -671,6 +671,10 @@ def __del__(self):
             finally:
                 self.__ray_terminate__._actor_hard_ref = None
 
+    def force_kill(self):
```
Comment (richardliaw) [inline python/ray/actor.py]: doc?

Comment (richardliaw) [reply python/ray/actor.py]: btw this is async right? any way to confirm actor death?

File: python/ray/tests/test_actor.py [hunk 4]
```diff
@@ -1431,6 +1431,22 @@ def ping(self):
...
+            # Never returns.
+            ray.get(ray.ObjectID.from_random())
+
+    actor = Actor.remote()
+    result = actor.hang.remote()
+    ready, _ = ray.wait([result], timeout=0.1)
+    assert len(ready) == 0
+    actor.force_kill()
+    with pytest.raises(ray.exceptions.RayActorError):
+        ray.get(result, timeout=1)
```
Comment (richardliaw) [inline python/ray/tests/test_actor.py]: how fast will `ray.available_resources` be updated?

**Review by richardliaw — COMMENTED** [review]
Looks good from cursory glance; a few questions:

1. what happens when the actor itself has a method called `force_kill`?
2. in general, should users prefer to use `__ray_terminate__` unless necessary?
3. Is `Shutdown()` equivalent to a sigterm?

Also, as @robertnishihara mentioned, can you make sure this is on `actors.rst` and the package reference? Thanks!

**Comment by jovany-wang** [comment]
It seems that Java have `exitActor()` API, but I didn't take into this PR.

Maybe we can unify something?

File: src/ray/core_worker/core_worker.cc (hunk 1 above)
Comment (edoakes) [reply src/ray/core_worker/core_worker.cc]: Added another log message that logs the actorID on creation

File: python/ray/actor.py (hunk 3 above)
Comment (edoakes) [reply python/ray/actor.py]: Yes it's async. You can wait on any outstanding actor tasks.

File: python/ray/tests/test_actor.py (hunk 4 above)
Comment (edoakes) [reply python/ray/tests/test_actor.py]: It's the same path as an actor/worker dying, so the raylet should update it very soon.

Comment (edoakes) [reply python/ray/tests/test_actor.py]: It's the same path as an actor/worker dying - the raylet should update it soon after death.

**Comment by edoakes** [comment]
@jovany-wang what are the semantics of the `exitActor()` API?

@richardliaw 
> 1. what happens when the actor itself has a method called `force_kill`?
You won't be able to call it. I'll change this to `__ray_kill__()` instead to avoid collisions.
> 2. in general, should users prefer to use `__ray_terminate__` unless necessary?
Yes this should be used for hanging. 
> 3. Is `Shutdown()` equivalent to a sigterm?
No. A user could install a sigterm handler that ignores the signal, but this won't be ignored. It's equivalent to stopping any existing python code and immediately running `sys.exit`.

File: src/ray/core_worker/core_worker.cc (hunk 1 above)
Comment (richardliaw) [reply src/ray/core_worker/core_worker.cc]: Sorry if this was unclear - I was more concerned about the situation where we couldn't trace which actor specifically was force killed. Does it make sense to log an ActorID with this "Shutting down" message?

**Comment by jovany-wang** [comment]
The semantics of `exitActor` and `__ray_kill__` are different I think.

An actor can invoke  `exitActor` to exit itself.

File: src/ray/core_worker/core_worker.cc (hunk 1 above)
Comment (edoakes) [reply src/ray/core_worker/core_worker.cc]: It's logged in the other log message and they both go to the logfile, not stderr.

**Comment by richardliaw** [comment]
The python API also has an `exit_actor()` API; I think it is different from this PR.

File: python/ray/actor.py [hunk 5]
```diff
@@ -671,6 +671,10 @@ def __del__(self):
             finally:
                 self.__ray_terminate__._actor_hard_ref = None
 
+    def __ray_kill__(self):
```
Comment (richardliaw) [inline python/ray/actor.py]: add docstring? and surface in `package-ref.rst`?

File: src/ray/core_worker/core_worker.cc (hunk 1 above)
Comment (richardliaw) [reply src/ray/core_worker/core_worker.cc]: Thanks, got it.

**Comment by richardliaw** [comment]
This PR looks good to me, but I am not familiar with this code - maybe @simon-mo or @ericl can take a brief look?

Also would be good to doc in package-ref and also add docstring.

**Review by richardliaw — APPROVED** [review]
Approving for now because it seems like an isolated change. Also, as mentioned, good to make sure it is in `package-ref.rst`.

Feel free to ask for another pass from someone else before merging.

**Review by simon-mo — APPROVED** [review]
yeah this look good to me as well. I like we can have ray terminate and ray kill to mirror sigterm and sigkill

File: python/ray/tests/test_actor.py [hunk 6]
```diff
@@ -1431,6 +1431,22 @@ def ping(self):
...
+    class Actor(object):
+        def hang(self):
+            # Never returns.
+            ray.get(ray.ObjectID.from_random())
+
+    actor = Actor.remote()
+    result = actor.hang.remote()
+    ready, _ = ray.wait([result], timeout=0.1)
+    assert len(ready) == 0
+    actor.force_kill()
```
Comment (raulchen) [inline python/ray/tests/test_actor.py]: I don't find where you inject this `force_kill` into the actor handle object. Did you forget that?

Comment (edoakes) [reply python/ray/tests/test_actor.py]: Ah sorry I renamed this but didn't update the test. It's now `__ray_kill__`

**Pull request merged**